src/main/java/com/iota/iri/hash/PearlDiver.java

```diff
@@ -29,230 +27,203 @@
     public void cancel() {
         synchronized (syncObj) {
             state = CANCELLED;
-            syncObj.notifyAll();
         }
     }
 
-    public synchronized boolean search(final int[] transactionTrits, final int minWeightMagnitude,
-        int numberOfThreads) {
-
+    private static void validateParameters(int[] transactionTrits, int minWeightMagnitude) {
         if (transactionTrits.length != TRANSACTION_LENGTH) {
             throw new RuntimeException(
                 "Invalid transaction trits length: " + transactionTrits.length);
         }
         if (minWeightMagnitude < 0 || minWeightMagnitude > CURL_HASH_LENGTH) {
             throw new RuntimeException("Invalid min weight magnitude: " + minWeightMagnitude);
         }
+    }
+
+    public synchronized boolean search(final int[] transactionTrits, final int minWeightMagnitude,
+                                       int numberOfThreads) {
 
+        validateParameters(transactionTrits, minWeightMagnitude);
         synchronized (syncObj) {
             state = RUNNING;
         }
 
-        final long[] midCurlStateLow = new long[CURL_STATE_LENGTH], midCurlStateHigh = new long[CURL_STATE_LENGTH];
-
-        {
-            for (int i = CURL_HASH_LENGTH; i < CURL_STATE_LENGTH; i++) {
-                midCurlStateLow[i] = HIGH_BITS;
-                midCurlStateHigh[i] = HIGH_BITS;
-            }
-
-            int offset = 0;
-            final long[] curlScratchpadLow = new long[CURL_STATE_LENGTH], curlScratchpadHigh = new long[CURL_STATE_LENGTH];
-            for (int i = (TRANSACTION_LENGTH - CURL_HASH_LENGTH) / CURL_HASH_LENGTH; i-- > 0; ) {
-
-                for (int j = 0; j < CURL_HASH_LENGTH; j++) {
-
-                    switch (transactionTrits[offset++]) {
-                        case 0: {
-                            midCurlStateLow[j] = HIGH_BITS;
-                            midCurlStateHigh[j] = HIGH_BITS;
+        final long[] midStateLow = new long[CURL_STATE_LENGTH];
+        final long[] midStateHigh = new long[CURL_STATE_LENGTH];
+        initializeMidCurlStates(transactionTrits, midStateLow, midStateHigh);
 
-                        }
-                        break;
-
-                        case 1: {
-                            midCurlStateLow[j] = LOW_BITS;
-                            midCurlStateHigh[j] = HIGH_BITS;
-                        }
-                        break;
-
-                        default: {
-                            midCurlStateLow[j] = HIGH_BITS;
-                            midCurlStateHigh[j] = LOW_BITS;
-                        }
-                    }
+        if (numberOfThreads <= 0) {
+            numberOfThreads = Math.max(1, Math.floorDiv(numberOfThreads * 8, 10));
```

Comment:
Won't this always set numberOfThreads to 1?

Reply:
OMG. THANKS FOR SPOTTING THAT!
see https://github.com/iotaledger/iri/pull/758

Reply:
I'm investigating now why the CI test didn't fail on: `https://github.com/iotaledger/iri-regression-tests/blob/master/pass_tx.py#L27`